### PR TITLE
[5.9] Deprecate Global Helper Functions

### PR DESCRIPTION
**Pull request by imliam**
**tldr**: Move global helper functions into a class, make the global functions just point to that class, if you don't want a function to be defined you can say so in your composer.json

# Deprecate Global Helper Functions

This PR is a is a preliminary step towards removing helper functions from the global scope, while not enforcing it on everyone yet.

It does this by moving the logic of all these functions into a new `\Laravel` class and making the helper functions simply be wrappers for these new methods.

Additionally, this offers a way for users of the framework to opt-in to whether they want the global helper functions to be defined, on their own terms.

The general motivation behind these changes is:

- To be _a step_ in the right direction
- Not to change the public API at all (yet)
- Allow users of the framework to solve the problems that global helpers can pose (which currently is not possible)

## Steps Taken

_Note: Each of these steps has been split out into separate commits, so you can easily see what changes were made as part of each one._

### 1. Register a `Laravel` class in the root namespace

The new class is where all the logic for the helper methods will be moved to.

The class is deliberately in the root namespace (outside `Illuminate`) to keep it short. This is to help alleviate the pains and concerns people will have when moving from a global function to the class variation - it's still short and sweet and there's no need for long import statements to use a helper method on-the-fly.

### 2. Move all foundation and support helper functions into the new class

The logic for these helpers has been split into corresponding traits in the support and foundation namespaces, which are then used in the `Laravel` class.

This means that a call to a helper function can be rewritten so that `app_path()` can now be used as `\Laravel::appPath()`.

If calling `app_path()` directly, it will itself call `\Laravel::appPath()` with the same arguments. This will allow all existing uses of the global functions to continue working as before.

### 3. Rewrite all calls to the global functions in the framework to the new way

The framework itself uses the global helper functions all over the place to make certain functionality work.

To allow the possible removal of global functions, these have all been changed to the new `Laravel` class method calls.

### 4. Remove the `function_exists` checks on global functions

The `function_exists` check on global functions prevents a runtime error if two functions are defined with the same name, which is quite often a _bad_ thing.

This means that if another function has been defined with the same name first, Laravel will quietly skip it and run a completely different function than you might expect. This can lead to unexpected behaviour that is hard to diagnose.

This has been evidenced several times in the past, such as in the [laravel/ideas/issues/1569](https://github.com/laravel/ideas/issues/1569) where the `hoa/protocol` package also defines a global `resolve` helper, causing Laravel's service container resolution to no longer work.

Removing the `function_exists` check will intentionally cause a runtime error that can be seen and dealt with immediately, instead of behaving unexpectedly.

### 5. Add an optional way for users to disable global helper functions

This is the real meat of the PR - the reason for all of the aforementioned changes to be in place.

The project's `composer.json` file can take an "extra" key with some additional properties for Laravel to read. If a `disabled-functions` key is found in here, its value will be read to determine if global helper functions should be defined or not.

If the value is set to `true`, it will prevent all global functions from being registered.

This puts it entirely on the end user to decide whether they want to have global helper functions available in their project or not.

```json
{
    "extra": {
        "laravel": {
            "disabled-functions": true
        }
    }
}
```

There is another option available too - passing an array of strings.

This will only disable the specified functions, leaving the rest of them completely untouched.

This will let the end user decide if they want to disable only a couple of functions that may be conflicting with another package while keeping the rest of the functions available for them.

```json
{
    "extra": {
        "laravel": {
            "disabled-functions": ["app", "resolve"]
        }
    }
}
```

## Next steps

Here are some additional miscellaneous thoughts I've had about where this can go from here:

- If this is adopted, all packages (both first-party and third-party) should update their usage of the global helper functions to the new methods. This will mean that if a user does disable any helpers, the packages will continue working as expected.

- The documentation could point to the new methods in the `Laravel` class instead of the global functions. This would start discouraging the use of global functions while not forcing it on anyone immediately.

- The global helper functions could be deprecated if that is the choice the Laravel team wants to take.

- We could implement a command to cache the enabled global helper functions to a single file to avoid the overhead of parsing the `composer.json` file on every request. Something like `php artisan helpers:cache`.

- The global helper functions could be moved to the `laravel/helpers` package.

- The string and array helpers in the `laravel/helpers` package could be wrapped in the same methods to be disabled on a case-by-case basis too.

- We could simply wrap the helper.php files in a namespace, but that will break backwards compatibility. It also seems pretty rare - I don't think I've ever seen namespaced functions in a package in the wild.

- This PR currently still has all the string and array support functions included, which were to be removed in 5.9

## Conclusion

I'm aware this approach isn't ideal, but it is meant to be a start to a discussion to move things in the right direction. Global functions don't _have_ to be deprecated, but they won't be a complete block for people using the framework.

I'm keen to hear everyone's thoughts and opinions on where to take it.

Edit: Oops I based this off the 5.8 branch instead of master. I'll sort that out depending on how the discussion goes.

### CODE REVIEW

**Comment by chriskonnertz** [comment]
> This PR is a is a preliminary step towards removing helper functions from the global scope

But there isn't a plan for further steps - or is there any? If there isn't a plan, then the deprecation anticipates something that is completely hidden in the shadows...

**Comment by driesvints** [comment]
I suggest you try pitching the idea first in the ideas repo before working on a pr.
